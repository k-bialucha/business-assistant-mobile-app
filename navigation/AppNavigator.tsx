import React from 'react';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { createStackNavigator } from '@react-navigation/stack';

import DashboardScreen from '../screens/DashboardScreen';
import SalesScreen from '../screens/SalesScreen';
import CostsScreen from '../screens/CostsScreen';
import CostsEntryScreen from '../screens/CostsEntryScreen';
import SalesEntryScreen from '../screens/SalesEntryScreen';

const DashboardStackNavigator = createStackNavigator();

const DashboardNavigator = () => {
  return (
    <DashboardStackNavigator.Navigator>
      <DashboardStackNavigator.Screen
        name="Dashboard"
        component={DashboardScreen}
      />
    </DashboardStackNavigator.Navigator>
  );
};

const SalesStackNavigator = createStackNavigator();

const SalesNavigator = () => {
  return (
    <SalesStackNavigator.Navigator>
      <SalesStackNavigator.Screen name="SalesList" component={SalesScreen} />
      <SalesStackNavigator.Screen
        name="SalesEntry"
        component={SalesEntryScreen}
      />
    </SalesStackNavigator.Navigator>
  );
};

const CostsStackNavigator = createStackNavigator();

const CostsNavigator = () => {
  return (
    <CostsStackNavigator.Navigator>
      <CostsStackNavigator.Screen name="CostsList" component={CostsScreen} />
      <CostsStackNavigator.Screen
        name="CostsEntry"
        component={CostsEntryScreen}
      />
    </CostsStackNavigator.Navigator>
  );
};

const AppTabNavigator = createBottomTabNavigator();

const AppNavigator = () => {
  return (
    <AppTabNavigator.Navigator>
      <AppTabNavigator.Screen name="Dashboard" component={DashboardNavigator} />
      <AppTabNavigator.Screen name="Sales" component={SalesNavigator} />
      <AppTabNavigator.Screen name="Costs" component={CostsNavigator} />
    </AppTabNavigator.Navigator>
  );
};

export default AppNavigator;
