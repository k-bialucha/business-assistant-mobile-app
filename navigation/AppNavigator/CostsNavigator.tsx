import React from 'react';

import { createStackNavigator } from '@react-navigation/stack';

import CostsEntryScreen from '../../screens/CostsEntryScreen';
import CostsScreen, { CostsScreenNavOptions } from '../../screens/CostsScreen';

import defaultScreenOptions from './defaultScreenOptions';
import { ScreenProps } from './index';

const CostsStackNavigator = createStackNavigator();

type Props = ScreenProps<'Costs'>;

const CostsNavigator: React.FC<Props> = () => {
  return (
    <CostsStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <CostsStackNavigator.Screen
        name="CostsList"
        component={CostsScreen}
        options={CostsScreenNavOptions}
      />
      <CostsStackNavigator.Screen
        name="CostsEntry"
        component={CostsEntryScreen}
      />
    </CostsStackNavigator.Navigator>
  );
};

export default CostsNavigator;
