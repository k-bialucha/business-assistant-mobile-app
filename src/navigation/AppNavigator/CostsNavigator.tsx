import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import CostsEntryScreen from '~/screens/CostsEntryScreen';
import CostsListScreen, {
  CostsListScreenNavOptions,
} from '~/screens/CostsListScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  CostsList: undefined;
  CostsEntry: undefined;
};

const CostsStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Costs'>;

const CostsNavigator: React.FC<Props> = () => {
  return (
    <CostsStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <CostsStackNavigator.Screen
        name="CostsList"
        component={CostsListScreen}
        options={CostsListScreenNavOptions}
      />
      <CostsStackNavigator.Screen
        name="CostsEntry"
        component={CostsEntryScreen}
      />
    </CostsStackNavigator.Navigator>
  );
};

export default CostsNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Costs'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
