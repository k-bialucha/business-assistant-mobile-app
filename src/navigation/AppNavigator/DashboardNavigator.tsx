import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import DashboardScreen from '~/screens/DashboardScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  Dashboard: undefined;
};

const DashboardStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Dashboard'>;

const DashboardNavigator: React.FC<Props> = () => {
  return (
    <DashboardStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <DashboardStackNavigator.Screen
        name="Dashboard"
        component={DashboardScreen}
      />
    </DashboardStackNavigator.Navigator>
  );
};

export default DashboardNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Dashboard'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
