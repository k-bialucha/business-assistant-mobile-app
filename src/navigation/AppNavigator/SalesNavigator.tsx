import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import SalesEntryScreen from '~/screens/SalesEntryScreen';
import SalesScreen, { SalesScreenNavOptions } from '~/screens/SalesScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  SalesList: undefined;
  SalesEntry: undefined;
};

const SalesStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Sales'>;

const SalesNavigator: React.FC<Props> = () => {
  return (
    <SalesStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <SalesStackNavigator.Screen
        name="SalesList"
        component={SalesScreen}
        options={SalesScreenNavOptions}
      />
      <SalesStackNavigator.Screen
        name="SalesEntry"
        component={SalesEntryScreen}
      />
    </SalesStackNavigator.Navigator>
  );
};

export default SalesNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Sales'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
