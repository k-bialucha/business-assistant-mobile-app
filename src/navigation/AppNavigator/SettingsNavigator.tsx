import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import SettingsScreen from '~/screens/SettingsScreen';
import UserSettingsScreen from '~/screens/UserSettingsScreen';
import { UserSettingsScreenNavOptions } from '~/screens/UserSettingsScreen/UserSettingsScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  Settings: undefined;
  UserSettings: undefined;
};

const SettingsStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Settings'>;

const SettingsNavigator: React.FC<Props> = () => {
  return (
    <SettingsStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <SettingsStackNavigator.Screen
        name="Settings"
        component={SettingsScreen}
      />
      <SettingsStackNavigator.Screen
        name="UserSettings"
        component={UserSettingsScreen}
        options={UserSettingsScreenNavOptions}
      />
    </SettingsStackNavigator.Navigator>
  );
};

export default SettingsNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Settings'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
