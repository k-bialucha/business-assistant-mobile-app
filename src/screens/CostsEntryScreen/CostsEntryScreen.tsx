import React from 'react';
import { ScrollView } from 'react-native';

import { Formik } from 'formik';
import { Button } from 'react-native-elements';
import RNPickerSelect from 'react-native-picker-select';
import * as Yup from 'yup';

import TextField from '~/components/form/TextField/TextField';
import { useAppTranslation } from '~/hooks/useAppTranslation';
import { AmountKind } from '~/models/AmountKind';
import { Currency } from '~/models/Currency';
import { VatRate, vatRates } from '~/models/VatRate';
import { NavigationData } from '~/navigation/AppNavigator/CostsNavigator';
import Colors from '~/theme/Colors';
import {
  ScreenMainStyledView,
  StyledWideContainer,
} from '~/theme/StyledComponents.styled';

type Props = NavigationData<'CostsEntry'>;

type CostFormItem = {
  amount: string;
  amountKind: AmountKind;
  currency: Currency | '';
  vatRate: VatRate | '';
  purchaseDate: string;
};

const CostsEntryScreen: React.FC<Props> = () => {
  const { t } = useAppTranslation();

  const CostEntrySchema = Yup.object<Partial<CostFormItem>>({
    amount: Yup.string().required(t('Required')),
    amountKind: Yup.string<AmountKind>().required(t('Required')),
    currency: Yup.string<Currency>().required(t('Required')),
    vatRate: Yup.string<VatRate>().required(t('Required')),
    purchaseDate: Yup.string().required(t('Required')),
  });
  const initialValues: CostFormItem = {
    amount: '',
    amountKind: AmountKind.GROSS, // 2x checkbox or switch
    // TODO: fill currency with base company currency in app?
    currency: '', // select field
    vatRate: '', // select field
    purchaseDate: '', // date field
  };

  return (
    <ScrollView
      keyboardShouldPersistTaps="handled"
      style={{ backgroundColor: '#ffffff' }}
    >
      <ScreenMainStyledView>
        <Formik
          initialValues={initialValues}
          validationSchema={CostEntrySchema}
          onSubmit={(cost: CostFormItem) => {
            // eslint-disable-next-line no-console
            console.log('Form sent! ', cost);
          }}
        >
          {({
            handleChange,
            handleBlur,
            handleSubmit,
            setFieldTouched,
            values,
            isValid,
            errors,
            touched,
          }) => (
            <StyledWideContainer>
              <TextField
                value={values.amount}
                placeholder={t('Amount')}
                onChangeText={handleChange('amount')}
                onBlur={handleBlur('amount')}
                error={errors.amount}
                touched={touched.amount}
                keyboardType="decimal-pad"
              />
              <TextField
                value={t(values.amountKind)}
                placeholder={t('Amount Kind')}
                onChangeText={handleChange('amountKind')}
                onBlur={handleBlur('amountKind')}
                error={errors.amountKind}
                touched={touched.amountKind}
              />
              <RNPickerSelect
                placeholder={{
                  label: t('Select a currency'),
                  value: '', // to override null and match Formik requirements
                  color: 'gray',
                }}
                onValueChange={handleChange('currency')}
                onClose={() => setFieldTouched('currency')} // imitating onBlur
                items={Object.values(Currency).map(item => ({
                  label: item,
                  value: item,
                }))}
              >
                <TextField
                  value={values.currency}
                  placeholder={t('Currency')}
                  error={errors.currency}
                  touched={touched.currency}
                />
              </RNPickerSelect>
              <RNPickerSelect
                placeholder={{
                  label: t('Select a Vat Rate'),
                  value: '', // to override null and match Formik requirements
                  color: 'gray',
                }}
                onValueChange={handleChange('vatRate')}
                onClose={() => setFieldTouched('vatRate')}
                items={vatRates.map(item => ({
                  label: `${item}%`,
                  value: item,
                }))}
              >
                <TextField
                  value={values.vatRate ? `${values.vatRate}%` : values.vatRate}
                  placeholder={t('Vat Rate')}
                  error={errors.vatRate}
                  touched={touched.vatRate}
                />
              </RNPickerSelect>
              <TextField
                value={values.purchaseDate}
                placeholder={t('Purchase Date')}
                onChangeText={handleChange('purchaseDate')}
                onBlur={handleBlur('purchaseDate')}
                error={errors.purchaseDate}
                touched={touched.purchaseDate}
              />
              <Button
                title={t('Save')}
                disabled={!isValid}
                titleStyle={{ color: Colors.silver }}
                onPress={() => {
                  handleSubmit();
                }}
              />
            </StyledWideContainer>
          )}
        </Formik>
      </ScreenMainStyledView>
    </ScrollView>
  );
};

export default CostsEntryScreen;
