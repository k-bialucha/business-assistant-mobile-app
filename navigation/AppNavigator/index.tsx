import React from 'react';
import { Platform } from 'react-native';

import { Ionicons } from '@expo/vector-icons';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';

import Colors from '../../theme/Colors';

import CostsNavigator from './CostsNavigator';
import DashboardNavigator from './DashboardNavigator';
import SalesNavigator from './SalesNavigator';
import SettingsNavigator from './SettingsNavigator';

export type RootStackParamList = {
  Dashboard: undefined;
  Sales: undefined;
  Costs: undefined;
  Settings: undefined;
};

export type ScreenProps<RouteName extends keyof RootStackParamList> = {
  navigation: StackNavigationProp<RootStackParamList, RouteName>;
  route: RouteProp<RootStackParamList, RouteName>;
};

const AppTabNavigator = createBottomTabNavigator<RootStackParamList>();

const AppNavigator: React.FC<{}> = () => {
  return (
    <AppTabNavigator.Navigator
      initialRouteName="Dashboard"
      screenOptions={({ route }) => ({
        tabBarIcon: ({ color, size }) => {
          const icons = {
            Dashboard: Platform.OS === 'android' ? 'md-stats' : 'ios-stats',
            Sales:
              Platform.OS === 'android' ? 'md-trending-up' : 'ios-trending-up',
            Costs:
              Platform.OS === 'android'
                ? 'md-trending-down'
                : 'ios-trending-down',
            Settings:
              Platform.OS === 'android' ? 'md-settings' : 'ios-settings',
          };

          return (
            <Ionicons name={icons[route.name]} size={size} color={color} />
          );
        },
      })}
      tabBarOptions={{
        showLabel: false,
        activeTintColor: Colors.navyBlue,
        inactiveTintColor: Colors.silver,
      }}
    >
      <AppTabNavigator.Screen name="Sales" component={SalesNavigator} />
      <AppTabNavigator.Screen name="Dashboard" component={DashboardNavigator} />
      <AppTabNavigator.Screen name="Costs" component={CostsNavigator} />
      <AppTabNavigator.Screen name="Settings" component={SettingsNavigator} />
    </AppTabNavigator.Navigator>
  );
};

export default AppNavigator;
