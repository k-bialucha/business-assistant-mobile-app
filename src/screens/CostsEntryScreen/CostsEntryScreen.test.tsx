import React from 'react';

import { shallow } from 'enzyme';

import { NavigationData } from '../../navigation/AppNavigator/CostsNavigator';

import CostsEntryScreen from './CostsEntryScreen';

type Props = NavigationData<'CostsEntry'>;

const fakeProps: Props = {
  route: { key: '1234', name: 'CostsEntry' },
  // @ts-ignore
  navigation: {},
};

describe('<CostsEntryScreen />', () => {
  it('matches the snapshot', () => {
    const tree = shallow(<CostsEntryScreen {...fakeProps} />);

    expect(tree).toMatchSnapshot();
  });
});
