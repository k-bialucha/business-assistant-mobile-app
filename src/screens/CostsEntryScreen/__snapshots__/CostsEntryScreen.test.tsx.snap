// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CostsEntryScreen /> matches the snapshot 1`] = `
<ScrollViewMock
  keyboardShouldPersistTaps="handled"
  style={
    Object {
      "backgroundColor": "#ffffff",
    }
  }
>
  <Styled(View)>
    <Formik
      initialValues={
        Object {
          "amount": "",
          "amountKind": "Net",
          "currency": "",
          "purchaseDate": 2021-01-01T00:00:00.000Z,
          "vatRate": "",
        }
      }
      onSubmit={[Function]}
      validationSchema={
        ObjectSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_defaultDefault": [Function],
          "_deps": Array [],
          "_excludedEdges": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_nodes": Array [
            "purchaseDate",
            "vatRate",
            "currency",
            "amountKind",
            "amount",
          ],
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "object",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "fields": Object {
            "amount": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "string",
            },
            "amountKind": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "string",
            },
            "currency": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "string",
            },
            "purchaseDate": DateSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "date",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "date",
            },
            "vatRate": NumberSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "number",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "number",
            },
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "object",
        }
      }
    >
      <Component />
    </Formik>
  </Styled(View)>
</ScrollViewMock>
`;
